### ISSUE TITLE
[no-floating-promises] Misleading documentation around what's "floating"

### ISSUE THREAD

**Issue by aeweda** [typescript-eslint/typescript-eslint#4589]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```TS
async function doSomething() {
    return 'something';
}

async function failEslintCorrectly() {
    doSomething();
}

async function passEslintIncorrectly() {
    const result = doSomething();
}
```

Rules:

```JSON
{
  "rules": {
    "react-hooks/rules-of-hooks": "error",
    "react-hooks/exhaustive-deps": "error",
    "@typescript-eslint/await-thenable": "error",
    "@typescript-eslint/no-misused-promises": "error",
    "@typescript-eslint/no-floating-promises": "error",
    "@typescript-eslint/restrict-template-expressions": "error",
    "@typescript-eslint/no-non-null-asserted-optional-chain": "error",
    "unused-imports/no-unused-imports": "error",
  }
}
```

we have the following rules set to off
```JSON
    "functional/prefer-readonly-type": "off",
    "@typescript-eslint/ban-types": "off",
    "@typescript-eslint/no-namespace": "off",
    "@typescript-eslint/ban-ts-comment": "off",
    "@typescript-eslint/no-unused-vars": "off",
    "@typescript-eslint/no-explicit-any": "off",
    "@typescript-eslint/no-var-requires": "off",
    "@typescript-eslint/no-empty-function": "off",
    "@typescript-eslint/no-empty-interface": "off",
    "@typescript-eslint/no-inferrable-types": "off",
    "@typescript-eslint/no-non-null-assertion": "off",
    "@typescript-eslint/explicit-module-boundary-types": "off",
    "yoda": "off",
    "curly": "off",
    "camelcase": "off",
    "brace-style": "off",
    "prefer-const": "off",
    "dot-notation": "off",
    "spaced-comment": "off",
    "prefer-rest-params": "off",
    "promise/param-names": "off",
    "prefer-regex-literals": "off",
    "node/handle-callback-err": "off",
    "lines-between-class-members": "off",
    "import/no-cycle": "off",
    "import/no-duplicates": "off",
    "import/no-absolute-path": "off",
    "import/no-named-default": "off",
    "import/newline-after-import": "off",
    "no-var": "off",
    "no-new": "off",
    "one-var": "off",
    "no-empty": "off",
    "no-sequences": "off",
    "no-undef-init": "off",
    "no-cond-assign": "off",
    "no-lone-blocks": "off",
    "no-fallthrough": "off",
    "no-self-compare": "off",
    "no-return-assign": "off",
    "no-useless-catch": "off",
    "no-useless-return": "off",
    "no-useless-escape": "off",
    "no-unneeded-ternary": "off",
    "no-case-declarations": "off",
    "no-use-before-define": "off",
    "no-unused-expressions": "off",
    "no-prototype-builtins": "off",
    "no-inner-declarations": "off",
    "no-constant-condition": "off",
    "no-useless-constructor": "off",
    "no-useless-computed-key": "off",
    "no-irregular-whitespace": "off"
```

**Expected Result**

it should fail in both cases

**Actual Result**

it passes when wrapped in a const

**Additional Info**

updating didn't work and nothing seems to get it to detect that there should be an await when wrapped in a const

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.12.1` |
| `@typescript-eslint/parser`        | `5.12.1` |
| `TypeScript`                       | `4.4.3` |
| `ESLint`                           | `8.9.0` |
| `node`                             | `14.17.0` |

**Comment by JoshuaKGoldberg** [comment]
@aeweda thanks for filing this issue!

The rule is working as intended. Sometimes it's useful to capture the result of creating a promise in a variable. In this case the `result` is type `Promise<string>`.

**Comment by burtonator** [comment]
@JoshuaKGoldberg 

> The rule is working as intended. Sometimes it's useful to capture the result of creating a promise in a variable. In this case the result is type Promise<string>.

If you leave the method, and don't await it , well, here be dragons.  

This is a HUGE bug and a major cause of problems on our end.

Is there another eslint plugin that does this? I mean if we're using the wrong tool thats fine. We'll use the right one.  but if there is NO tool then this is definitely an eslint bug. 

Note that you can also bypass this by passing it as a method variable to another method.

**Comment by JoshuaKGoldberg** [comment]
What would you propose be done to prevent this bug? Banning variables whose types are `Promise<...>`? Enforce that if a Promise variable is created, it must either be returned, awaited, or `.then`-d?

**Comment by bradzacher** [comment]
IMO the rule shouldn't ever flag this case because it's not a floating promise - it's an explicitly assigned promise.

The bug you're looking at here is that you have an unused variable, which has captured the promise. By leaving an unused variable in the codebase, the value is never properly handled.

There's a rule for that issue - [`no-unused-vars`](https://typescript-eslint.io/rules/no-unused-vars). If you enforce that there's no unused variables in your codebase then you'll ensure that this class of bugs can't occur.

**Comment by burtonator** [comment]
@JoshuaKGoldberg @bradzacher thanks for the feedback guys! My thoughts below.

> What would you propose be done to prevent this bug? Banning variables whose types are Promise<...>? Enforce that if a Promise variable is created, it must either be returned, awaited, or .then-d?

I think that unless it's returned, the promise must either have a catch or an await.  

One other issue. IF we don't fix this in this rule, or if it can't be fixed, we should have a BIG callout/warning in the docs about this issue.

We had assumed this was the behavior and I'm worried other people would too and it might bite them.

I get your point about these not being 'floating' as they're assigned but I think the intent is to not lose any promises like this and not await/catch them because that's an amazing source of errors.

Nearly ALL our promise errors are around this issue and we really want to crush them. 

When you have a bigger team you're tryin to scale you need tools like this to spot these issues sooner.

> IMO the rule shouldn't ever flag this case because it's not a floating promise - it's an explicitly assigned promise.

This might be a bit pedantic though.

The goal is to not have any broken promise code and the behavior on node/browser is different and we've seen subtle implementation behavior that's weird across promise implementations.

This is absolutely the #1 cause of promise related bugs for us.

> The bug you're looking at here is that you have an unused variable, which has captured the promise. By leaving an unused variable in the codebase, the value is never properly handled.

We're using it... we're passing it to another function.

That isn't caught either :-/ Even if it's not named or assigned.

> There's a rule for that issue - [no-unused-vars](https://typescript-eslint.io/rules/no-unused-vars). If you enforce that there's no unused variables in your codebase then you'll ensure that this class of bugs can't occur.

It's used.  We're passing it to another function but that function has the the type 'any' so we try to JSON encode it.

This is just a technicality though as this is an entire / larger class of bugs for us.  We MUST make sure a promise is awaited/caught.

> Rules are purposely limited in scope and focused. If we reported on this variable then it would mean the rule overlaps with no-unused-vars, which is not correct.

The issue isn't related here.

The method exits without that promise being awaited or caught and that's a huge problem.  

Sorry if I feel very passionate about this issue.  As I mentioned it's by far our #1 promise related but and I'd like to kill it once and for all :)

**Comment by JoshuaKGoldberg** [comment]
> callout/warning in the docs about this issue.

Indeed. Accepting PRs to rephrase https://typescript-eslint.io/rules/no-floating-promises ([no-floating-promises.md](https://github.com/typescript-eslint/typescript-eslint/blob/04baac8d8958a21d9ae3312206dc283fec642887/packages/eslint-plugin/docs/rules/no-floating-promises.md)).

I'm not sold that a huge warning banner is called for; it seems to me like the rule description just doesn't describe what "floating" means in the rule's context.

---

I see what you're saying that storing promises in variables not handled appropriately can lead to them being floating. It is unfortunate that the rule does not try to stop that, but doing so would be quite a lot of work. Consider the following code snippet:

```ts
declare const startLongTask: () => Promise<void>;

// Happy case: handlesDirectly directly handling variable ✔

function handlesDirectly() {
  const pendingResult = startLongTask();

  pendingResult
    .then(() => console.log("Done!"))
    .catch((error) => console.error(error));
}

// Happy case: delegatesProperly passes to a proper function ✔

function processPendingResult(pendingResult: Promise<void>) {
  pendingResult
    .then(() => console.log("Done!"))
    .catch((error) => console.error(error));
}

function delegatesProperly() {
  const pendingResult = startLongTask();

  processPendingResult(pendingResult);
}

// Unhappy case: delegatesImproperly provides to function that doesn't handle it ✖

function logUnknown(data: unknown) {
    console.log(`[log] ${data}`);
}

function delegatesImproperly() {
  const pendingResult = startLongTask();

  logUnknown(pendingResult);
  console.log(pendingResult);
}
```

How could a lint rule mark the difference between `delegatesProperly` passing a `Promise` value to a function that handles it appropriately vs. `delegatesImproperly` passing a `Promise` value to a function that doesn't? Two directions come to mind:

* Banning storing `Promise` values _at all_ 
    * Would you also ban `logUnknown(startLongTask())`?
* Preventing passing `Promise` to locations typed as `any` or `unknown`
    * Some of these are caught by [no-misused-promises](https://typescript-eslint.io/rules/no-misused-promises) at the moment

For banning storing `Promise` values _at all_, that could be a legitimate strategy for some codebases but is far too opinionated at the moment for typescript-eslint core. You could write your own ESLint rules using the APIs described in [Custom Rules](https://typescript-eslint.io/docs/development/custom-rules).

For `no-misused-promises`, if there are additional cases you'd like the rule to catch, I'd request you file >=1 new issue(s) with specific suggestions. This one is getting a bit long. 😄

**Comment by bradzacher** [comment]
To give you the full explanation as to why this is entirely working as intended.

All static analysis tooling (TS and ESLint are exactly the same in this regard) do not and cannot analyse across function or module boundaries. It's possible to build but incredibly difficult for such little gain.

Which means that there is absolutely no way for us to know whether or not the function that you passed your variable to handled the promise appropriately. Which leaves us with two choices: assume it did or assume it didn't.

We must assume that it did, or else there is so, so much valid code that would false-positive that it would train users to ignore the rule and miss the true positives.

As a very simple example:

```ts
const x = Promise.resolve();
await Promise.race([X]);
```

This code is 100% valid, and clearly so - but how can we know that statically it did use the promise correctly? If we reported on it then everyone would not use the rule.

----

When a promise is assigned to a variable we assume that you are going to use the variable (every codebase I've seen has the recommended rule `no-unused-vars` or the tsconfig option `noUnusedLocals`, so that's a pretty safe assumption in the general case).

When a promise is passed to a function, we (have to) assume that function correctly handles the promise because there's no way to analyse across the boundary.

When a promise is returned or yielded we assume the caller correctly handles the promise. If it doesn't it will be caught by `no-floating-promise` within that function or some cases will be caught by `no-misused-promises`. For return there's also `return-await` which can enforce awaiting a returned promise if you want.

In weird positions like in an `if` condition we assume TS will catch it (as it does for some cases), and for all other cases are delegated to the rule `no-misused-promises`.

All that is left is the "expression statements" - which is what `no-floating-promises` handles.

**Comment by intellix** [comment]
This one just screwed me over royally as well. A payment method: **isValidSignature** changed from sync to async and suddenly they weren't being checked. Thanks for mentioning **no-misused-promises** because that actually solves it for me. Would be nice to get recommendations when you use something like **no-floating-promises** because I was also falsly assuming I was safe... think it might have been caught in tslint before I migrated?

**Comment by JoshuaKGoldberg** [comment]
> think it might have been caught in tslint before I migrated?

No, those cases aren't checked by the old rule. https://github.com/palantir/tslint/blob/master/test%2Frules%2Fno-floating-promises%2Fpromises%2Ftest.ts.lint

At any rate, I threw up a documentation PR at #4649 and would appreciate any thoughts from folks! 💖
